src/V3HierBlock.cpp

```diff
@@ -0,0 +1,420 @@
+
+string V3HierBlock::hierLib(bool withDir) const { return hierSomeFile(withDir, "lib", ".a"); }
+
+string V3HierBlock::hierGenerated(bool withDir) const {
+    return hierWrapper(withDir) + ' ' + hierMk(withDir);
+}
+
+void V3HierBlock::writeCommandFile(bool forCMake) const {
+    vl_unique_ptr<std::ofstream> of(V3File::new_ofstream(commandFileName(forCMake)));
+    *of << "--cc\n";
+
+    if (!forCMake) {
+        for (V3HierBlock::HierBlockSet::const_iterator child = m_children.begin();
+             child != m_children.end(); ++child) {
+            *of << v3Global.opt.makeDir() << "/" << (*child)->hierWrapper(true) << "\n";
+        }
+    }
+    *of << "-Mdir " << v3Global.opt.makeDir() << "/" << hierPrefix() << " \n";
+    writeCommonInputs(of.get(), forCMake);
+    const V3StringList& commandOpts = commandOptions(false);
+    for (V3StringList::const_iterator it = commandOpts.begin(); it != commandOpts.end(); ++it) {
+        *of << (*it) << "\n";
+    }
+    *of << hierBlockOptions().front() << "\n";
+    for (HierBlockSet::const_iterator child = m_children.begin(); child != m_children.end();
+         ++child) {
+        *of << hierBlockOptions().front() << "\n";
+    }
+    *of << v3Global.opt.allArgsStringForHierBlock(false) << "\n";
+}
+
+string V3HierBlock::commandFileName(bool forCMake) const {
+    return hierCommandFileName(hierPrefix(), forCMake);
+}
+
+//######################################################################
+// Collect how hierarchy blocks are used
+class HierBlockUsageCollectVisitor : public AstNVisitor {
+    // NODE STATE
+    // AstNode::user1()            -> bool. Processed
+    AstUser1InUse m_inuser1;
+
+    // STATE
+    typedef std::set<const AstModule*> ModuleSet;
+    V3HierBlockPlan* const m_planp;
+    AstModule* m_modp;
+    ModuleSet m_referred;  // Modules that have hier_block pragma
+    V3HierBlock::GParams m_gparams;  // list of variables that is AstVarType::GPARAM
+
+    virtual void visit(AstModule* nodep) VL_OVERRIDE {
+        // Don't visit twice
+        if (nodep->user1SetOnce()) return;
+        UINFO(5, "Checking " << nodep->prettyNameQ() << " from "
+                             << (m_modp ? m_modp->prettyNameQ() : string("null")) << std::endl);
+        AstModule* const prevModulep = m_modp;
+        ModuleSet prevReferred;
+        V3HierBlock::GParams prevGParams;
+        if (nodep->hierBlock()) {
+            m_modp = nodep;
+            prevReferred.swap(m_referred);
+        }
+        prevGParams.swap(m_gparams);
+
+        iterateChildren(nodep);
+
+        if (nodep->hierBlock()) {
+            m_planp->add(nodep, m_gparams);
+            for (ModuleSet::const_iterator it = m_referred.begin(); it != m_referred.end(); ++it) {
+                m_planp->registerUsage(nodep, *it);
+            }
+            m_modp = prevModulep;
+            m_referred.swap(prevReferred);
+        }
+        m_gparams.swap(prevGParams);
```

Comment:
Why swap?  Just assign?  Ditto other places.

Reply:
swap is cheaper than assignment. (swap is expected to exchange just an internal pointer)
C++11 offers move semantics and would be more intuitive.
```c++
m_gparams = std::move(prevGParams);
```

Reply:
It can only swap a pointer with variables that are local (directly or inlined) to the function, it won't work with items on the heap as can't change other functions.

I believe most recent compilers are pretty good at promoting assignment to std::move, but haven't benchmarked it.

Reply:
swap of STL containers (e.g. map, set, vector) are fast regardless of whether the container itself is on heap or stack.
(contents of the container must be on heap, of course. So std::array::swap() is slow).

https://github.com/gcc-mirror/gcc/blob/5abe0d81db068c06ae57b011e1a9d4fd7976ac56/libstdc%2B%2B-v3/include/bits/stl_vector.h#L1469-L1473
https://github.com/gcc-mirror/gcc/blob/5abe0d81db068c06ae57b011e1a9d4fd7976ac56/libstdc%2B%2B-v3/include/bits/stl_set.h#L431-L434

Reply:
But yes, I agree that we can expect compiler to do a good job.
will use assignment.

Reply:
Assignment is used in 7c06ed2e32daf1025fe6cf934f372a464d19237a.